### DISCUSSION TITLE
GPU scalar memory is part of global memory or not?

### DISCUSSION THREAD

**Discussion by ShuangqiLi** [gem5/gem5#1796] [Q&A] [answered]
https://github.com/gem5/gem5/blob/b1a44b89c7bae73fae2dc547bc1f871452075b85/src/gpu-compute/scoreboard_check_stage.cc#L224

the function isGlobalMem() is defined at
https://github.com/gem5/gem5/blob/b1a44b89c7bae73fae2dc547bc1f871452075b85/src/gpu-compute/gpu_static_inst.hh#L178

The code in these two places seems contradictory. If scalar memory is part of global memory, then the function should OR (||) with _flags[Scalar] during the check; otherwise, the logic in the scoreboard_check_stage.cc would be redundant.

**Comment by abmerop** [comment, marked as answer]
This part of scoreboard is checking if there is a functional unit available to issue the instruction. A scalar address is *usually* backed by global memory (vram/framebuffer), but it will use the scalar load unit which skips coalescer and is routed through the scalar cache.

You could think about the code this way if that makes more sense:

```
    } else if (ii->isMemRef()) {
        if (!ii->isScalar()) {
            return w->globalMem;
        } else {
            return w->scalarMem;
        }
```

**Reply by ShuangqiLi** [reply]
Got it, thank you!
